### ISSUE TITLE
search site hardcoded to specfic host

### ISSUE THREAD

**Issue by coke** [Raku/doc-website#128]
on docs-dev, if you search for something and select "search entire site", you get sent to the old dev site:

https://www.google.com/search?q=site%3Anew-raku.finanalyst.org+plugins

**Comment by coke** [comment]
This needs to pulled out into a config or env. var so it can be easily set by the CI.

**Comment by finanalyst** [comment]
what am I missing?

**Comment by coke** [comment]
... a config that isn't checked into the repository.

**Comment by finanalyst** [comment]
OK. Please let me know what the best way to do this should be. Are these variable that a Raku program could see using %*ENV?

**Comment by finanalyst** [comment]
More information. When a web-page is rendered, a search-bar.js is generated. The default 'found' item is a call to Google's search pointing at the href of the website (eg., new-raku.finananalyst.org, or docs.raku.org) and the chars in the search input field.
The href is currently taken from `Website/configs/03-plugin-options.raku` and the key `search-bar->search-site`.
If the href of the website is fixed, then this config setting can just be set (hard-coded as explained by @coke )
If the href of the website has to be according to an environment variable, that is possible too. For example if RAKU_WEB_SITE (just to illustrate) is set in the environment, then its value can be used for the Google command. 
But it will be the same value as the one where users go to find the site in the first place, so it is rather fixed.

**Comment by coke** [comment]
For go live, we can change the hardcoded value to 'docs.raku.org' which avoids the issue. After go-live, we can make it configurable during build time so that when running on docs-dev or raku-new it works seamlessly.

**Comment by finanalyst** [comment]
@coke making the search item automatically configurable for developers will come after we get a set of steps for a local set up. I think this needs some feedback from other people. 
For Go-Live, I can change the config now to docs.raku.org (I'll create a PR for this now).
Then we can change the tooling according to specific requests of users. eg. an environment variable, an option to build-site, or leave as is.
So lets change the status of this issue to Cleanup.

**Comment by dontlaugh** [comment]
My preference is the following:

* Allow parameterizing this via a command line flag at build time. Call the parameter `base-url` or `base_url`. This is a common requirement for static site generators. [Example from Zola](https://www.getzola.org/documentation/getting-started/configuration/). Do not support a default. Fail the build unless the script is provided explicitly.
* Do not add a new script. Do not add new library dependencies.
* For interactive dev server use case (e.g. a local Cro server), allow - perhaps - a localhost default, and support the same flag to set it explicitly.

**Comment by coke** [comment]
> My preference is the following:
> * For interactive dev server use case (e.g. a local Cro server), allow - perhaps - a localhost default, and support the same flag to set it explicitly.

localhost makes no sense - google has nothing to search here; default here should be the production site. Otherwise I agree with you.

**Comment by dontlaugh** [comment]
No, we need parameters for this. If it's not easy to parameterize, that's something we need to fix.

**Comment by dontlaugh** [comment]
But fwiw it's not really that urgent, right?

**Comment by finanalyst** [comment]
@dontlaugh I'm being dumb here. What do you mean by parameterize. I don't think there is a problem to fix, so much as I dont understand what you want fixing.

Its not urgent, but I want to close issues as soon as possible. I hated Raku/docs with its thousands of issues.

**Comment by coke** [comment]
parameterize - replaced hardcoded value with a variable, or parameter.

Yes, this needs fixing; the deployment step has control over the deployed hostname, and this needs to match. 

Yes, this is not urgent.

No, we're not closing non-urgent items that need to be in the backlog. You can skip items in the issue listing that don't have a milestone, that should help.

**Comment by finanalyst** [comment]
@coke I am still trying to understand what the problem is here, and what sort of solution would be acceptable. I'm not trying to insist on any point of view, just to understand the principle.
I looked at the Zola link that @dontlaugh provided above. But I cannot see how - in principle - that example differs from what we already have.

**Comment by coke** [comment]
We deploy the same container to multiply named hosts (prod & non-prod). We may eventually have to change the hostname, as we've done in the past with doc/docs.

The deployed code currently has a *checked in* copy of the hostname. This means that to override it, we'd have to
* change a file under version control (not ideal), or:
* ignore the hostname and always search the production site (not preferred in general but ok as a fallback)

So to address these concerns, we're looking for a way to pass in the name of the host we're deploying to to the script and use that to generate the site.

I'm assuming the zola example is not *checked into the repo*

**Comment by finanalyst** [comment]
@coke I've been thinking about this all night :)

I think you want the following:
- create a container P
- deploy the same container P on one host (called prod) that serves on url docs.raku.org
  - the search default points to `docs.raku.org`
- deploy the same container P on another host (called dev) that serves on url docs-dev.raku.org
  - the search default points to `docs-dev.raku.org`

From search of docs, it seems possible to deploy the container eg for the `dev` host as
```
docker run -e 'SEARCH_URL=doc-dev.raku.org' <other container run options>
```
This **I think** creates an ENV variable which a Raku program running inside the container can pick up as `%*ENV<SEARCH_URL>`

But this does not resolve the issue because the search-bar is driven by a JS script which is generated as a static asset at the time the container is built, not when it is run.

At present, all assets are served by NGINX, which is running inside the container, as static files. 

Suppose, instead that we run a Cro App inside the container and behind NGINX? The Cro APP serves all routes except for the asset `search-bar.js` as static, but `search-bar.js` is created by the Cro App with the search default entity set to the value of the `%*ENV<SEARCH_URL>` 

In addition, there is the suggestion that `search-bar.js` runs as a microservice getting data as requested, rather than having all the data bundled into the static asset. In this case, the Cro App could then be configured so the `search-bar.js` is dynamically created, and it contains a Websocket address that is also dynamically created to point back at the host where the container is deployed to.

Is this an approach to this issue? If yes, I can work on it in due course

**Comment by 2colours** [comment]
Is the main issue that the JS search has a hardcoded URL? If so, that could be dynamically addressed by JS itself, using the `window.location` API, and generating the value of that into the query.

**Comment by finanalyst** [comment]
@2colours aha!!!! much better idea. I'll look into that as soon as I have time

**Comment by 2colours** [comment]
More accurately: If we change [this line](https://github.com/Raku/doc-website/blob/0df605c4b616cbbdf9f5e383e4e8c4ad092c40a8/Website/plugins/search-bar/add-search.raku#L89) to contain `window.location.host` instead, the search would always happen on the host itself. Of course this could be refined in a way that allows for a fallback on the JS side, e.g "if our host is localhost, let's fall back to the predefined hostname" or something like that.

**Comment by finanalyst** [comment]
the fix in #170 solves this problem completely. No need for any server side change.
